File: src/chapter04/Problem21_BOJ1517.java
package chapter04;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Problem21_BOJ1517 {
    public static int[] arr;
    public static int[] temp;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        arr = new int[N];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        mergeSort(0, arr.length - 1);
    }

    private static void mergeSort(int start, int end) {
        if (end - start < 1) {
            return;
        }

        int mid = (start + end) / 2;
        mergeSort(start, mid);
        mergeSort(mid + 1, end);

        int index = start;
        int i = start;
        int j = mid + 1;

        while (i <= mid && j <= end) {

        }


    }
}

File: src/chapter04/Problem21_BOJ1517.py
import sys
input = sys.stdin.readline

n = int(input())
arr = list(map(int, input().split()))
temp = list
answer = 0

def mergeSort(start, end):
    global answer, arr

    if (end - start < 1):
        return
    
    mid = (end + start) // 2
    mergeSort(start, mid)
    mergeSort(mid + 1, end)

    index = start
    i = start
    j = mid + 1
    temp = []

    while (i <= mid and j <= end):
        if (arr[i] > arr[j]):
            temp.append(arr[j])
            answer += (mid - i) + 1
            j += 1
        else:
            temp.append(arr[i])
            i += 1

    if (i <= mid):
        temp += arr[i:mid+1]
    if (j <= end):
        temp += arr[j:end+1]
    
    for i in range(len(temp)):
        arr[start+i] = temp[i]

mergeSort(0, len(arr) - 1)
print(answer)